### PR TITLE
Extended proxy_(de)serialize objects to work with stringstream, ostringstream, istringstream and vector<char>

### PR DESCRIPTION
**Pull request by pfeatherstone**

### CODE REVIEW

File: dlib/serialize.h [hunk 1]
```diff
@@ -1655,34 +1707,31 @@ namespace dlib
                 if (objects_read == 0)
                 {
                     throw serialization_error("An error occurred while trying to read the first" 
-                        " object from the file " + filename + ".\nERROR: " + e.info + "\n" + suffix);
+                        " object from the stream.\nERROR: " + e.info + "\n" + suffix);
```
Comment (davisking) [inline dlib/serialize.h:1710]: Definitely don't remove the filename.  That's a hugely important part of the error message.

File: dlib/serialize.h [hunk 2]
```diff
@@ -1583,47 +1584,86 @@ namespace dlib
...
-            // file.
-            fin->read(file_header,4);
-            fin->clear();
-            fin->seekg(0);
+            init();
         }
-
+        
+        explicit proxy_deserialize (
+            std::istringstream& ss
```
Comment (davisking) [inline dlib/serialize.h:1644]: Why not just make this `std::istream`?

File: dlib/serialize.h [hunk 3]
```diff
@@ -1583,47 +1584,86 @@ namespace dlib
...
+        ) : fin_optional_owning_ptr(nullptr),
+            fin(ss)
+        {
+            init();
+        }
+        
+        explicit proxy_deserialize (
+            std::stringstream& ss
```
Comment (davisking) [inline dlib/serialize.h:1652]: Same here.  This isn't needed if the constructor above just takes `std::istream`

**Comment by davisking** [comment]
Looks good, aside from the few things I noted.

File: dlib/serialize.h (hunk 2 above)
Comment (pfeatherstone) [reply dlib/serialize.h:1644]: Yes good point.

File: dlib/serialize.h (hunk 3 above)
Comment (pfeatherstone) [reply dlib/serialize.h:1652]: Yep

**Review by davisking — CHANGES_REQUESTED** [review]

**Comment by davisking** [comment]
Looks good, thanks for another PR :)

**Pull request merged**